src/lib/hash/sha3/keccak_fips.cpp

```diff
@@ -0,0 +1,208 @@
+         {
+         S[S_pos / 8] ^= load_le<uint64_t>(input, 0);
+         S_pos += 8;
+         input += 8;
+         to_take -= 8;
+         }
+
+      while(to_take)
+         {
+         S[S_pos / 8] ^= static_cast<uint64_t>(input[0]) << (8 * (S_pos % 8));
+
+         ++S_pos;
+         ++input;
+         --to_take;
+         }
+
+      if(S_pos == bitrate / 8)
+         {
+         Keccak_FIPS::permute(S.data());
+         S_pos = 0;
+         }
+      }
+
+   return S_pos;
+   }
+
+//static
+
+void Keccak_FIPS::finish(size_t bitrate,
+                                 secure_vector<uint64_t>& S, size_t S_pos, uint64_t custom_padd, uint8_t custom_padd_bit_len)
+   {
+   BOTAN_ARG_CHECK(bitrate % 64 == 0, "Keccak-FIPS bitrate must be multiple of 64");
+   // append the first bit of the final padding after the custom padding
+   uint8_t init_pad = custom_padd | (1 << custom_padd_bit_len);
+   S[S_pos / 8] ^= static_cast<uint64_t>(init_pad) << (8 * (S_pos % 8));
+   // final bit of the padding of the last block
+   S[(bitrate / 64) - 1] ^= static_cast<uint64_t>(0x80) << 56;
+   Keccak_FIPS::permute(S.data());
+   }
+
+//static
+
+void Keccak_FIPS::expand(size_t bitrate,
+                   secure_vector<uint64_t>& S,
+                   uint8_t output[], size_t output_length)
+   {
+   BOTAN_ARG_CHECK(bitrate % 64 == 0, "Keccak-FIPS bitrate must be multiple of 64");
+
+   const size_t byterate = bitrate / 8;
+
+   while(output_length > 0)
+      {
+      const size_t copying = std::min(byterate, output_length);
+
+      copy_out_vec_le(output, copying, S);
+
+      output += copying;
+      output_length -= copying;
+
+      if(output_length > 0)
+         {
+         Keccak_FIPS::permute(S.data());
+         }
+      }
+   }
+
+
+Keccak_FIPS::Keccak_FIPS(std::string const& base_name, size_t output_bits, size_t capacity,
+      uint64_t custom_padd,
+      uint8_t custom_padd_bit_len) :
+   m_output_bits(output_bits),
+   m_bitrate(1600 - capacity),
+   m_capacity(capacity),
+   m_custom_padd(custom_padd),
+   m_custom_padd_bit_len(custom_padd_bit_len),
+   m_base_name(base_name),
+   m_S(25), // 1600 bit
+   m_S_pos(0)
+   {
+   // We only support the parameters for Keccak-FIPS in this constructor
+
+   if(output_bits > 1600 )
+      throw Invalid_Argument("Keccak_FIPS: Invalid output length " +
+                             std::to_string(output_bits));
+   }
+
+
+std::string Keccak_FIPS::name() const
+   {
+   return m_base_name + "(" + std::to_string(m_output_bits) + ")";
+   }
+
+
+std::string Keccak_FIPS::provider() const
+   {
+#if defined(BOTAN_HAS_KECCKAK_FIPS_BMI2)
+   if(CPUID::has_bmi2())
+      {
+      return "bmi2";
+      }
+#endif
+
+   return "base";
+   }
+
+
+void Keccak_FIPS::clear()
+   {
+   zeroise(m_S);
+   m_S_pos = 0;
+   }
+
+
+void Keccak_FIPS::add_data(const uint8_t input[], size_t length)
+   {
+   m_S_pos = Keccak_FIPS::absorb(m_bitrate, m_S, m_S_pos, input, length);
+   }
```

Comment:
Two things:

1. We don't need to stick to the `HashFunction` interface here. Let's consider to adopt the Keccak-nomenclature and call this `absorb` (just not static)
2. For new interfaces, let's use `std::span<>` consistently

```suggestion
void Keccak_FIPS::absorb(std::span<const uint8_t> input)
   {
   m_S_pos = Keccak_FIPS::absorb(m_bitrate, m_S, m_S_pos, input.data(), length.size());
   }
```

Usage of the interface will need adaption, of course.

Reply:
Done in upcoming commit.